Source/Core/VideoCommon/XFStructs.cpp

```diff
@@ -300,3 +301,431 @@ void PreprocessIndexedXF(u32 val, int refarray)
+    switch (chan.diffusefunc)
+    {
+    case LIGHTDIF_NONE:
+      diffusefunc = "None";
+      break;
+    case LIGHTDIF_SIGN:
+      diffusefunc = "Sign";
+      break;
+    case LIGHTDIF_CLAMP:
+      diffusefunc = "Clamp";
+      break;
+    }
+    switch (chan.attnfunc)
+    {
+    case LIGHTATTN_NONE:
+      attnfunc = "No attenuation";
+      break;
+    case LIGHTATTN_SPEC:
+      attnfunc = "Point light attenuation";
+      break;
+    case LIGHTATTN_DIR:
+      attnfunc = "Directional light attenuation";
+      break;
+    case LIGHTATTN_SPOT:
+      attnfunc = "Spot light attenuation";
+      break;
+    }
+    return format_to(ctx.out(),
+                     "Material source: {0}\nEnable lighting: {1}\nLight mask: {2:x} ({2:08b})\n"
+                     "Ambient source: {3}\nDiffuse function: {4}\nAttenuation function: {5}",
+                     chan.matsource ? "Yes" : "No", chan.enablelighting ? "Yes" : "No",
+                     chan.GetFullLightMask(), chan.ambsource ? "Yes" : "No", diffusefunc, attnfunc);
+  }
+};
+
+template <>
+struct fmt::formatter<TexMtxInfo>
+{
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const TexMtxInfo& i, FormatContext& ctx)
+  {
+    std::string_view texgentype = "Invalid", sourcerow = "Invalid";
+    // This is a 3-bit field for some reason, despite listed values only using 2 bits
+    switch (i.texgentype)
+    {
+    case XF_TEXGEN_REGULAR:
+      texgentype = "Regular";
+      break;
+    case XF_TEXGEN_EMBOSS_MAP:
+      texgentype = "Emboss map (used when bump mapping)";
+      break;
+    case XF_TEXGEN_COLOR_STRGBC0:
+      texgentype = "Color STRGBC0";
+      break;
+    case XF_TEXGEN_COLOR_STRGBC1:
+      texgentype = "Color STRGBC1";
+      break;
+    }
+    switch (i.sourcerow)
+    {
+    case XF_SRCGEOM_INROW:
+      sourcerow = "Geometry (input is abc)";
+      break;
+    case XF_SRCNORMAL_INROW:
+      sourcerow = "Normal (input is abc)";
+      break;
+    case XF_SRCCOLORS_INROW:
+      sourcerow = "Colors";
+      break;
+    case XF_SRCBINORMAL_T_INROW:
+      sourcerow = "Binormal T (input is abc)";
+      break;
+    case XF_SRCBINORMAL_B_INROW:
+      sourcerow = "Binormal B (input is abc)";
+      break;
+    case XF_SRCTEX0_INROW:
+      sourcerow = "Tex 0";
+      break;
+    case XF_SRCTEX1_INROW:
+      sourcerow = "Tex 1";
+      break;
+    case XF_SRCTEX2_INROW:
+      sourcerow = "Tex 2";
+      break;
+    case XF_SRCTEX3_INROW:
+      sourcerow = "Tex 3";
+      break;
+    case XF_SRCTEX4_INROW:
+      sourcerow = "Tex 4";
+      break;
+    case XF_SRCTEX5_INROW:
+      sourcerow = "Tex 5";
+      break;
+    case XF_SRCTEX6_INROW:
+      sourcerow = "Tex 6";
+      break;
+    case XF_SRCTEX7_INROW:
+      sourcerow = "Tex 7";
+      break;
+    }
+    return format_to(ctx.out(),
+                     "Projection: {}\nInput form: {}\nTex gen type: {}\n"
```

Comment:
Should this be size instead of 'projection'?  I think of different things when I hear 'projection' but maybe I'm misunderstanding

Reply:
YAGCD calls it projection, while libogc's [`GX_SetTexCoordGen`](https://libogc.devkitpro.org/gx_8h.html#a7d3139b693ace5587c3224e7df2d8245) calls it `tgen_typ` and has [`GX_TG_MTX3x4` and `GX_TG_MTX2x4`](https://libogc.devkitpro.org/group__texgentyp.html#ga37b44995ede4932a201bc8edc2578c9e); the 3 by 4 version divides by `Q` which I think is the actual projection step.  But I don't think calling it projection is particularly obvious (not is typ(e)); size might be better.